backend/dataall/base/utils/iam_policy_utils.py

```diff
@@ -0,0 +1,95 @@
+    logger.info(f"Number of statements = {len(statements)}")
+    logger.info(f"Total length of statements = {total_length}")
+    max_length = max(statements_list_of_strings, key=len)
+    if len(max_length) > POLICY_LIMIT - POLICY_HEADERS_BUFFER:
+        raise Exception(f"Policy statement {max_length} exceeds maximum policy size")
+    while index < len(statements):
+        chunk = []
+        chunk_size = 0
+        while index < len(statements) and chunk_size + len(str(statements[index].to_json())) < POLICY_LIMIT - POLICY_HEADERS_BUFFER:
+            chunk.append(statements[index])
+            chunk_size += len(str(statements[index].to_json()))
+            index += 1
+        chunks.append(chunk)
+    logger.info(f"Total number of managed policies = {len(chunks)}")
+    if len(chunks) > MAXIMUM_NUMBER_MANAGED_POLICIES:
+        raise Exception("The number of policies calculated exceeds the allowed maximum number of managed policies")
+    return chunks
+
+
+def split_policy_with_resources_in_statements(base_sid, effect, actions, resources):
+    """
+    Splitter used for IAM policy statements with an undefined number of resources.
+    - Ensures that the size of the IAM statement is below the POLICY LIMIT
+    - If it exceeds the POLICY LIMIT, it breaks the statement in multiple statements with a subset of resources
+    - Note the POLICY_HEADERS_BUFFER to account for the headers of the policy which usually take around ~60chars
+    """
+    statement_without_resources = iam.PolicyStatement(
+        sid=base_sid,
+        effect=effect,
+        actions=actions,
+        resources=["*"]
+    )
+    resources_str = '" ," '.join(r for r in resources)
+    number_resources = len(resources)
+    max_length = len(max(resources, key=len))
+    base_length = len(str(statement_without_resources.to_json()))
+    total_length = base_length + len(resources_str)
+    logger.info(f"Policy base length = {base_length}")
+    logger.info(f"Number of resources = {number_resources}, resource maximum length = {max_length}")
+    logger.info(f"Resources as string length = {len(resources_str)}")
+    logger.info(f"Total length approximated as base length + resources string length = {total_length}")
+
+    if total_length < POLICY_LIMIT - POLICY_HEADERS_BUFFER:
+        logger.info("Not exceeding policy limit, returning statement ...")
+        resulting_statement = iam.PolicyStatement(
+            sid=base_sid,
+            effect=effect,
+            actions=actions,
+            resources=resources
+        )
+        return [resulting_statement]
+    else:
+        logger.info("Exceeding policy limit, splitting statement ...")
+        index = 0
+        split = 0
+        resulting_statements = []
+        while index < len(resources):
+            size = 0
+            res = []
+            while index < len(resources) and (size + len(resources[index]) + 5) < POLICY_LIMIT - POLICY_HEADERS_BUFFER - base_length:
```

Comment:
what is the `+5` here? An additional buffer for resource list?

Reply:
It accounts for the additional 4 chars around a resource string. Having a list of resources ["res1", "res2", "res3"]. If we only account the chars of each of the items then we would be counting only res1res2res3. The `", "` between resources are 4 chars. Then at the beginning and end we have the extra " and [], that's why I added an extra +1 and rounded up to 5. I added a comment for this

Reply:
Makes sense - I think I was seeing `resources_str` include those additional characters for the `total_length` of the policy and makes sense then  checking for each resource to add to the statement